### DISCUSSION TITLE
thread/synchronized

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#241] [Announcements]
# thread/synchronized

Java中的每一个对象都可以作为一个锁，这是synchronized实现同步的基础。当我们调用一个用synchronized关键字修饰的方法时，我们需要获取这个方法所在对象的锁。只有获取了这个锁，才可以执行这个方法。如果锁已经被其他线程获取，那么就会进入阻塞状态，直到锁被释放。

https://javabetter.cn/thread/synchronized.html

**Comment by Norwa9** [comment]
线程栈中Lock Record是什么？

**Reply by mking2023** [reply]
Lock Record，指锁记录，线程获取锁后会在栈中创建一个锁记录，表示线程获取到了锁，锁记录中包含锁的标识符等信息。线程释放锁后，锁记录会被删除

**Comment by Ricehahaa** [comment]
这个锁升级真的好难受啊，网上好多细节都不太一样

**Comment by zisoqlsdha** [comment]
轻量锁发生锁竞争时，采用自适应自旋策略，自适应自旋如果成功，则增加下一次自旋的次数，如果失败则减少下一次自旋的次数。但问题是，自旋失败了，锁就升级为重量锁了，重量锁的竞争又没有自旋这个操作，那减少下一次自旋次数的意义是啥？

**Reply by lin667z** [reply]
锁可能会降级，优化可能再次发生的轻量级锁竞争行为（同一个锁对象）。​
